pkg/ipamd/ipamd.go

```diff
@@ -538,37 +553,25 @@ func (c *IPAMContext) nodeInit() error {
 		vpcV4CIDRs = c.updateCIDRsRulesOnChange(vpcV4CIDRs)
 	}, 30*time.Second)
 
-	eniConfigName, err := eniconfig.GetNodeSpecificENIConfigName(ctx, c.cachedK8SClient)
-	if err == nil && c.useCustomNetworking && eniConfigName != "default" {
-		// Signal to VPC Resource Controller that the node is using custom networking
-		err := c.SetNodeLabel(ctx, vpcENIConfigLabel, eniConfigName)
+	if c.enablePodENI {
+		// call GET labels only once to avoid duplicate the call from labelling trunk and eniconfig
+		labels, err := c.getNodeLabels(ctx)
 		if err != nil {
-			log.Errorf("Failed to set eniConfig node label", err)
-			podENIErrInc("nodeInit")
 			return err
 		}
-	} else {
-		// Remove the custom networking label
-		err := c.SetNodeLabel(ctx, vpcENIConfigLabel, "")
-		if err != nil {
-			log.Errorf("Failed to delete eniConfig node label", err)
-			podENIErrInc("nodeInit")
-			return err
+
+		// if node labels is nil in rare cases, we will log the info instead of failing VPC CNI startup
+		if labels == nil {
+			// if labels is nil, ipamd logs the warning for debugging
+			log.Warn("node labels are nil during ipamd starts up, may need investigation before using the feature of security group for pods")
 		}
-	}
 
-	if metadataResult.TrunkENI != "" {
-		// Signal to VPC Resource Controller that the node has a trunk already
-		err := c.SetNodeLabel(ctx, "vpc.amazonaws.com/has-trunk-attached", "true")
-		if err != nil {
-			log.Errorf("Failed to set node label", err)
-			podENIErrInc("nodeInit")
-			// If this fails, we probably can't talk to the API server. Let the pod restart
-			return err
+		if c.useCustomNetworking {
+			// not doing retry here, don't have to handle error since error has been logged in the called method
+			c.askForENIConfigIfNeeded(ctx, labels)
 		}
-	} else {
-		// Check if we want to ask for one
-		c.askForTrunkENIIfNeeded(ctx)
+		// we want to ask VPC RC once during node init to cover special cases
+		c.askForTrunkENIIfNeeded(ctx, labels)
```

Comment:
Do we skip adding events for unsupported instances? I didn't see a check or I might have missed during code walk thru.

Reply:
We won't. VPC resource controller will check if the instance is supported. Since the instance type is checked from a static file in resource controller and so far we don't have a API to check during runtime, VPC CNI can't check and send event.

Reply:
Yes. I will update README and also check if that can be feasible to use a dummy interface on unsupported instance types.